src/perf_mainloop.cc

```diff
@@ -144,29 +201,51 @@ static void worker(DDProfContext *ctx, const WorkerAttr *attr,
 
         // Attempt to dispatch the event
         struct perf_event_header *hdr = rb_seek(rb, tail);
-        DDRes res = ddprof_worker(hdr, pes[i].pos, &flags->can_run, ctx);
+        DDRes res = ddprof_worker_process_event(hdr, pes[i].pos, ctx);
 
         // We've processed the current event, so we can advance the ringbuffer
         rb->region->data_tail += hdr->size;
 
         // Check for processing error
         if (IsDDResNotOK(res)) {
-          attr->finish_fun(ctx);
-          WORKER_SHUTDOWN();
+          return res;
         } else {
           if (hdr->type == PERF_RECORD_SAMPLE)
             ++processed_samples;
         }
       }
-    } while (events);
 
-    // If I didn't process any events, then hit the timeout
-    if (!processed_samples) {
-      if (IsDDResNotOK(ddprof_worker_timeout(&flags->can_run, ctx))) {
-        attr->finish_fun(ctx);
-        WORKER_SHUTDOWN();
+      DDRES_CHECK_FWD(ddprof_worker_maybe_export(restart_worker, ctx));
```

Comment:
The way you've consolidated the export logic is certainly much cleaner.  I think the changes in this area are fine.  In particular, checking for `!processed_samples` doesn't seem like a meaningful check in the current codebase.

At one point we had a sleep test.  As long as that passes, this should be good.

Reply:
No, this is not good, you are right. I forgot to add back the export check in case of `poll` timeout.